RecoLocalMuon/CSCSegment/src/CSCSegFit.h

```diff
@@ -0,0 +1,127 @@
+#ifndef CSCSegment_CSCSegFit_h
+#define CSCSegment_CSCSegFit_h
+
+// CSCSegFit.h  - Segment fitting factored out of CSC segment builders - Tim Cox
+// Last mod: 03.02.2015
+
+/* This as an object which is initialized by a set of rechits (2 to 6) in a 
+ * specific CSC and has the functionality to make a least squares fit to a 
+ * straight line in 2-dim for those rechits.
+ * The covariance matrix and chi2 of the fit are calculated.
+ * The original code made use of CLHEP matrices but this version uses 
+ * ROOT SMatrices because they are  multithreading compatible.
+ * Because of this, the no. of rechits that can be handled is limited to
+ * a maximum of 6, one per layer of a CSC. This means maximum dimensions
+ * can be specified at compile time and hence satisfies SMatrix constraints.
+ * For 2 hits of course there is no fit - just draw a straight line between them.
+ * Details of the algorithm are in the .cc file
+ *
+ */
+   
+#include <DataFormats/CSCRecHit/interface/CSCRecHit2D.h>
+#include <Geometry/CSCGeometry/interface/CSCChamber.h>
+
+#include <Math/Functions.h>
+#include <Math/SVector.h>
+#include <Math/SMatrix.h>
+
+#include <vector>
+
+class CSCSegFit {
+
+public:
+
+// TYPES
+
+   typedef std::vector<const CSCRecHit2D*> CSCSetOfHits;
+   
+  // 12 x12 Symmetric
+  typedef ROOT::Math::SMatrix<double,12,12,ROOT::Math::MatRepSym<double,12> > SMatrixSym12;
+
+  // 12 x 4
+  typedef ROOT::Math::SMatrix<double,12,4 > SMatrix12by4;
+
+  // 4 x 4 General + Symmetric
+  typedef ROOT::Math::SMatrix<double, 4 > SMatrix4;
+  typedef ROOT::Math::SMatrix<double,4,4,ROOT::Math::MatRepSym<double,4> > SMatrixSym4;
+
+  // 2 x 2 Symmetric
+  typedef ROOT::Math::SMatrix<double,2,2,ROOT::Math::MatRepSym<double,2> > SMatrixSym2;
+
+  // 4-dim vector
+  typedef ROOT::Math::SVector<double,4> SVector4;
+
+
+  // PUBLIC FUNCTIONS
+
+  //@@ WANT OBJECT TO CACHE THE SET OF HITS SO CANNOT PASS BY REF
+  CSCSegFit( const CSCChamber* csc, CSCSetOfHits hits) : 
+  chamber_( csc ), hits_( hits ), scaleXError_( 1.0 ), fitdone_( false ) {}
+
+  virtual ~CSCSegFit() {}
+
+  // Least-squares fit
+  void fit( void ); // fill uslope_, vslope_, intercept_  @@ FKA fitSlopes()
+  // Calculate covariance matrix of fitted parameters
+  AlgebraicSymMatrix covarianceMatrix(void);
+
+  // Change scale factor of rechit x error 
+  // - expert use only!
+  void setScaleXError ( double factor ) { scaleXError_ = factor; }
+
+  // Fit values
+  float xfit( float z ) const;
+  float yfit( float z ) const;
+
+  // Deviations from fit for given input (local w.r.t. chamber)
+  float xdev( float x, float z ) const;
+  float ydev ( float y, float z ) const;
+  float Rdev( float x, float y, float z ) const;
+
+  // Other public functions are accessors
+  CSCSetOfHits hits(void) const { return hits_; }
```

Comment:
Hi Slava,

The original code in several of the algorithms uses the set of hits as a cache and often manipulates it, inserting or erasing elements. My simple approach was just to make sure the CSCSegFit could satisfy their requirements, so I wanted a copy of the initial set of hits held in the CSCSegFit object. 

I am not sure the existing code will work even with a reference, let alone a const reference. Maybe it can be changed, but that is an optimization for the future, not a show stopper, and somebody would have to ensure everything continues to work fine.

On a side note, Piet and I were discussing earlier in the week how to adapt this for GEM too, since the same CLHEP-based code has also been entangled in the GEM local reco. It is conceivable that the code could be extended to TrackingRecHits, and dimensions will have to be increased to at least allow 8 hits. At the same time one could attempt to clarify ownership and constness of the set of hits. 

Regards, Tim

> On Feb 21, 2015, at 00:05 , Slava Krutelyov <notifications@github.com> wrote:
> 
> In RecoLocalMuon/CSCSegment/src/CSCSegFit.h:
> 
> > +
> > +  // Change scale factor of rechit x error 
> > +  // - expert use only!
> > +  void setScaleXError ( double factor ) { scaleXError_ = factor; }
> > +
> > +  // Fit values
> > +  float xfit( float z ) const;
> > +  float yfit( float z ) const;
> > +
> > +  // Deviations from fit for given input (local w.r.t. chamber)
> > +  float xdev( float x, float z ) const;
> > +  float ydev ( float y, float z ) const;
> > +  float Rdev( float x, float y, float z ) const;
> > +
> > +  // Other public functions are accessors
> > +  CSCSetOfHits hits(void) const { return hits_; }
> 
> why not return a const CSCSetOfHits& ?
> 
> —
> Reply to this email directly or view it on GitHub.
>